RecoEgamma/EgammaElectronProducers/plugins/SiStripElectronSeedProducer.h

```diff
@@ -12,15 +12,15 @@
 
 class SiStripElectronSeedGenerator;
 
-class SiStripElectronSeedProducer : public edm::EDProducer
+class SiStripElectronSeedProducer : public edm::global::EDProducer<>

```

Comment:
I wonder whether this producer can remain a ```global``` producer, or should it better become a ```stream``` one.

My doubt comes from:
https://github.com/guitargeek/cmssw/blob/11892b2b28f50b2488cda7ad1f1084f67b243c7c/RecoEgamma/EgammaElectronProducers/plugins/SiStripElectronSeedProducer.cc#L80
where the ```matcher_```  class member gets updated in the ```produce()``` method.

If I look at what the ```setupES(iSetup)``` method of ```SiStripElectronSeedGenerator``` does
https://github.com/guitargeek/cmssw/blob/11892b2b28f50b2488cda7ad1f1084f67b243c7c/RecoEgamma/EgammaElectronAlgos/src/SiStripElectronSeedGenerator.cc#L99
there is the theoretical possibility that the cached members of ```SiStripElectronSeedGenerator``` which are checked therein are different than the ones taken from the setup, and therefore get reset independently in every stream. 

If this is actually the case, then this producer must become a ```stream``` one, instead.

@Dr15Jones : may I ask your opinion about it?

Reply:
I'm sure you are right there! I did this PR when I was not yet aware of the multithreading subtleties. I was very naive and thought once a global module compiles it's fine because otherwise there will be complaints that `const` is not respected.

I guess here the problem is that `matcher_` is a pointer, and the pointer is not changed so the const of `produce` is respected. I did some study on that. The following class compiles:

```
// A counter class
class C {
  public:
    C() : c_(0) {}
    void increment() { ++c_; }
  private:
    int c_;
};

// Class which has a pointer to a counter as member
class A {
  public:
    A() : m_(new C) {}
    ~A() { delete m_; }
    void produce () const { m_->increment(); }
  private:
    C * m_;
};
```

It wouldn't compile with `C const * m_`. Even better would be `C const * const m_` becuase no stream should just point the pointer to another `const C`. So I guess the only was to make sure a global module will behave well is to verify that all members are either const or const pointers to const objects (same for references I guess), right? As this is not possible for the `SiStripElectronSeedProducer`, it should become a stream module. I should change that.